### ISSUE TITLE
After upgrading from 2.0.20 to 4.0.1, Unity crashes debilitatingly frequently

### ISSUE THREAD

**Issue by CodeHatchling** [ValveSoftware/steam-audio#169]
I've upgraded to 4.0.1 from 2.0.20. I'm using Unity 2020.2.0f1.

The editor crashes extremely often now. After hitting the [>] Play button to test, there's about a 1/3 chance that it will crash.

The crash logs reveal that these crashes are always from within a Steam Audio related DLL, usually Phonon:

1. 'Attempt to access invalid address.' ... 0x00007FFFCE08D16D (audioplugin_phonon) UnityGetAudioEffectDefinitions
2. Read from location FFFFFFFFFFFFFFFC caused an access violation. ... 0x00007FFF829B5420 (phonon) embree::TaskScheduler::wait
3. Read from location FFFFFFFFFFFFFFFF caused an access violation. ... 0x00007FFF829B5420 (phonon) embree::TaskScheduler::wait
4. 'Attempt to access invalid address.' ... (audioplugin_phonon) UnityGetAudioEffectDefinitions
5. 'Attempt to access invalid address.' ... (audioplugin_phonon) UnityGetAudioEffectDefinitions
6. 'Attempt to access invalid address.' ... (audioplugin_phonon) UnityGetAudioEffectDefinitions
7. 'Attempt to access invalid address.' ... (audioplugin_phonon) UnityGetAudioEffectDefinitions
8. Read from location 0000002000000011 caused an access violation. ... (phonon) embree::TaskScheduler::wait
9. Read from location 0000002000000011 caused an access violation. ... (phonon) embree::TaskScheduler::wait

I uploaded the crash logs here.
[Crash_2021-09-22_020218439.zip](https://github.com/ValveSoftware/steam-audio/files/7225588/Crash_2021-09-22_020218439.zip)

The scene most of these crashes occurred in involves a simple cube-shaped room with baked pathing and reverb. Probes are placed using custom code using the ProbeBatch and ProbeArray classes, though these crashes occur in levels without these things added.

**Comment by CodeHatchling** [comment]
Something worth noting about the debug logs: I logged the values pointed to by IntPtr instances in case they were using bad values, in hopes of "catching" these problems in C# where bad references do not destroy the universe.

**Comment by CodeHatchling** [comment]
Interested in knowing an ETA for when this issue can be looked into. Crashing is constant, might have to revert to 2.0.20.

**Comment by lakulish** [comment]
Thanks for providing the crash dumps, they all seem to indicate crashes when looking up baked pathing data. A couple of questions regarding the crashing behavior you're experiencing:

* Does the crash only occur soon after (or within a couple of seconds of) clicking Play? Or does it also occur after being in Play mode for several seconds or more?
* Can you provide a minimal scene that reproduces the crashes? We are attempting to repro this anyway, but if you could provide a scene or project that we can use, it'll help us.

Thanks!

**Comment by CodeHatchling** [comment]
- Sometimes it crashes about a second after pressing play (or any steam audio will become silent)
- Sometimes it crashes after about 5-10 seconds after pressing play.
- Sometimes it crashes after playing for a few minutes or so.
- Sometimes it displays a error popup window with a circled red X that closes Unity when you press OK on the window. While this message is present it is hard to tell what components of SteamAudio are still working. (These crashes do not produce a Crash log in Local/Temp/Unity/Crashes)
- Some of the crashes may be attributed to SteamOverlay (esp. the ones that don't produce crahs logs), but most of the ones I've checked were from Steam Audio.

**Comment by CodeHatchling** [comment]
I'll try to send you a repro scene.

However, if for some reason I cannot reproduce it (perhaps I made an unusual mistake during installation) another option could be to create a sort of "safe mode" branch in which the sore spots in code are checked for unexpected values e.g. null pointers, which could help diagnose the issue.

**Comment by CodeHatchling** [comment]
Another point worth mentioning is these crashes occur in levels without pathing data. That is, none of the sources are set to use pathing, and no pathing data is baked into any probe groups.

**Comment by lakulish** [comment]
Can you send a couple of crash dumps from when Unity crashes on a scene with no pathing data? Most of the crash dumps you previously sent indicate a crash in pathing code (and the rest don't contain enough information for us to know where the crash occurred). Note that Unity's logs may be inaccurate in reporting where the crash occurred, the crash dumps themselves are more useful.

Also, does the issue occur in a fresh Unity project? You mentioned that you upgraded from 2.0-beta.20 to 4.0.1, so I'm wondering if something went wrong in this process.

**Comment by CodeHatchling** [comment]
Good news!

I managed to get the crash reproducing in a minimal project, however I cannot upload it here as it is 242 MB.
But when you do get it, open the project in Unity 2020.2.0f1.

However, it appears that the crashes depend on another element and won't trigger by itself.

I'm pooling audio sources:
- A fixed number of audio sources are generated initially.
- When they play, the mixer, audio clip, and other properties are assigned before the sound is told to play.
- When more sounds attempt to play than the number of pooled audio sources, it estimates which audio source is the least audible and recycles it.

There are multiple ways to crash the demo, and shouldn't take long.

### General repro steps, in any order:

- Open the level **Assets\Scenes\Sandbox\Sandbox_Room_Small.unity** and press Play.
- Duplicate the cube named "Metal Cube". Spam them if you need to. Move them around. The idea is to trigger collision sound effects.
- Move the camera (listener) around, outside of the level.
- Wait. Sometimes it happens when you're doing nothing.
- Start and stop playing repeatedly, especially after doing the above steps in between.

### Specific repro steps:
**Runtime Error Window Crash**
1. Open and play the level "Sandbox_Room_Small" and move the camera outside of the cube. (e.g. position 100, 100, 100)
2. A popup box will instantly appear: "Microsoft Visual C++ Runtime Library - Runtime Error! Program: ..."
![ErrorWindow](https://user-images.githubusercontent.com/7232874/135442804-37d27bf2-013b-48d0-a39f-4083c85567b3.png)
3. If you press [OK] Unity will shut down, but will seem to be running fine in the background if you don't touch the window instead.
4. It'll probably crash if you wait long enough regardless of what you do.

**Wait a second Crash**
1. Open and play the level "Sandbox_Room_Small"
2. Wait. It'll probably crash in under 10 seconds or so.

**Spam Crash**
1. Open and play the level "Sandbox_Room_Small"
2. Duplicate "Metal Cube" over and over. Move them around in the scene. Make noise.
3. Wait a minute after they settle, it'll eventually crash.

**Comment by CodeHatchling** [comment]
[Crash_2021-09-30_105224284.zip](https://github.com/ValveSoftware/steam-audio/files/7259049/Crash_2021-09-30_105224284.zip)

Here are some more crash dumps.

The following crash occurred with the pathing data being removed, and all audio sources having the "pathing" boolean disabled.
[CrashNoPathing.zip](https://github.com/ValveSoftware/steam-audio/files/7259056/CrashNoPathing.zip)

**Comment by CodeHatchling** [comment]
I'm trying to simplify the conditions to repro the crash, and have managed to remove the requirement to have sounds triggered by physics. Instead, it bombards it with requests to play audio for a given arbitrary collision condition.

Strangely, I had a similar script running in another duplicate project that just randomly told a fixed number of audio sources to play multiple times per frame at different positions, with different properties (pitch, volume, range) and after letting it run for a half an hour it didn't crash.

But for the former case it seems to crash when the level of activity rises and then drops to nothing...

If this delineation between crashing vs. stable conditions is useful I can continue peeling away layers until I find the simplest repro steps. Then again, if you've already repro'ed and figured out what it is, that'd be great.

**Comment by ArgentumPhoenix** [comment]
Same issue. Unity 2021.1.22f1, SteamAudio 4.0.1. UnityEditor crashes. It happens at different time.
[Crash_2021-10-03_145924542.zip](https://github.com/ValveSoftware/steam-audio/files/7273647/Crash_2021-10-03_145924542.zip)

**Comment by CodeHatchling** [comment]
Unfortunately I'm going to have to revert to 2.0.20, at least temporarily. The team requires the sound to be functional on the main branch. An ETA on the fixes for these crashes would be greatly appreciated!

**Comment by ctorkut** [comment]
I can confirm that these crashes are not exclusive to the Editor, but they occur in builds with the exact same stack trace as well.

**Comment by lakulish** [comment]
@CodeHatchling Thanks for providing the test project! That helped us identify two crashes: one that applies to sources with pathing enabled, and another that applies to any source. Fixes for these issues will be included in the next patch release.

**Comment by CodeHatchling** [comment]
Thank you so much for fixing these crashes! :D
If possible, could you provide a rough time that the next patch can be expected? Thanks either way.

I'll close the issue when I confirm the crashes have stopped.

**Comment by DenisSulyagin** [comment]
Obtained 37 stack frames
0x00007ffb6219d012 (audioplugin_phonon) UnityGetAudioEffectDefinitions
0x00007ff7a17fbf71 (Unity) UnityAudioEffect_InternalSetFloatParameterCallback
0x00007ff7a30491be (Unity) FMOD::DSP::setParameter
0x00007ff7a1826335 (Unity) AudioSource::SetSpatializerFloat
0x00007ff7a078e73d (Unity) AudioSource_CUSTOM_SetSpatializerFloat
0x000002daa9a7d5f9 (Mono JIT Code) (wrapper managed-to-native) UnityEngine.AudioSource:SetSpatializerFloat (UnityEngine.AudioSource,int,single)
0x000002daa9a7d89b (Mono JIT Code) [UnityAudioEngineSource.cs:94] SteamAudio.UnityAudioEngineSource:SetSpatializerIntPtr (intptr,int&) 
0x000002daa9a7d38b (Mono JIT Code) [UnityAudioEngineSource.cs:56] SteamAudio.UnityAudioEngineSource:UpdateParameters (SteamAudio.SteamAudioSource) 
0x000002daa9a7c201 (Mono JIT Code) [SteamAudioSource.cs:166] SteamAudio.SteamAudioSource:Awake () 
0x000002db7bbe3ec8 (Mono JIT Code) (wrapper runtime-invoke) object:runtime_invoke_void__this__ (object,intptr,intptr,intptr)
0x00007ffb21b17ed0 (mono-2.0-bdwgc) [mini-runtime.c:3445] mono_jit_runtime_invoke 
0x00007ffb21a596c4 (mono-2.0-bdwgc) [object.c:3064] do_runtime_invoke 
0x00007ffb21a5985c (mono-2.0-bdwgc) [object.c:3111] mono_runtime_invoke 
0x00007ff7a1333ff4 (Unity) scripting_method_invoke
0x00007ff7a132ec94 (Unity) ScriptingInvocation::Invoke
0x00007ff7a132ed7e (Unity) ScriptingInvocation::InvokeChecked
0x00007ff7a13a81a6 (Unity) SerializableManagedRef::CallMethod
0x00007ff7a12fc4b4 (Unity) MonoBehaviour::CallAwake
0x00007ff7a12fad75 (Unity) MonoBehaviour::AddToManager
0x00007ff7a12fb6f1 (Unity) MonoBehaviour::AwakeFromLoad
0x00007ff7a13e91df (Unity) AwakeFromLoadQueue::InvokePersistentManagerAwake
0x00007ff7a13e9c4c (Unity) AwakeFromLoadQueue::PersistentManagerAwakeFromLoad
0x00007ff7a13e9da3 (Unity) AwakeFromLoadQueue::PersistentManagerAwakeFromLoad
0x00007ff7a1046542 (Unity) LoadSceneOperation::CompleteAwakeSequence
0x00007ff7a1046fa2 (Unity) LoadSceneOperation::CompletePreloadManagerLoadSceneEditor
0x00007ff7a1047b3f (Unity) LoadSceneOperation::IntegrateMainThread
0x00007ff7a104b701 (Unity) PreloadManager::UpdatePreloadingSingleStep
0x00007ff7a104bcdf (Unity) PreloadManager::WaitForAllAsyncOperationsToComplete
0x00007ff7a21f019e (Unity) EditorSceneManager::RestoreSceneBackups
0x00007ff7a1f1329f (Unity) PlayerLoopController::EnterPlayMode
0x00007ff7a1f223aa (Unity) PlayerLoopController::SetIsPlaying
0x00007ff7a1f253a2 (Unity) Application::TickTimer
0x00007ff7a235a9e5 (Unity) MainMessageLoop
0x00007ff7a235f1db (Unity) WinMain
0x00007ff7a365f482 (Unity) __scrt_common_main_seh
0x00007ffb9d477034 (KERNEL32) BaseThreadInitThunk
0x00007ffb9dfc2651 (ntdll) RtlUserThreadStart

**Comment by CodeHatchling** [comment]
I haven't experienced crashes since updating, so it is an improvement! Apparently there are still crashes though, so I wouldn't say this one is closed yet.

**Comment by kisak-valve** [comment]
Closing as fixed due to a lack of a response from @DenisSulyagin.

**Comment by themars2011** [comment]
This still happens on the newest version. We are on 2021.2.8f1

**Comment by kisak-valve** [comment]
Hello @themars2011, please open a new issue report with the details of the crash you are seeing.

**Comment by Vasco-Rosa** [comment]
Hello, has there been any solutions for this problem?

I am having a similar issue, where every time I try to generate probes my unity crashes. I am using Unity 2020.3.26f1 and Steam Audio 4.0.3. 

The report log is the following:

Obtained 54 stack frames
0x00007fff52949be9 (phonon) embree::TaskScheduler::wait
0x00007fff51819e5a (phonon) dummy_rtcore_ispc___sse2
0x00007fff518199f2 (phonon) dummy_rtcore_ispc___sse2
0x00007fff51819069 (phonon) dummy_rtcore_ispc___sse2
0x00007fff5294c76e (phonon) embree::TaskScheduler::wait
0x00007fff5296d86a (phonon) iplContextRetain
0x00007fff5296dc66 (phonon) iplContextRetain
0x000001e5d1bfb729 (Mono JIT Code) (wrapper managed-to-native) SteamAudio.API:iplStaticMeshLoad (intptr,intptr,SteamAudio.ProgressCallback,intptr,intptr&)
0x000001e5d1bfaea3 (Mono JIT Code) [StaticMesh.cs:71] SteamAudio.StaticMesh:.ctor (SteamAudio.Context,SteamAudio.Scene,SteamAudio.SerializedData) 
0x000001e5d1bf693b (Mono JIT Code) [SteamAudioProbeBatch.cs:135] SteamAudio.SteamAudioProbeBatch:GenerateProbes () 
0x000001e5d1bdc0db (Mono JIT Code) [SteamAudioProbeBatchInspector.cs:53] SteamAudio.SteamAudioProbeBatchInspector:OnInspectorGUI () 
0x000001e5d1bc4424 (Mono JIT Code) UnityEditor.UIElements.InspectorElement/<>c__DisplayClass59_0:<CreateIMGUIInspectorFromEditor>b__0 ()
0x000001e4e323bc0f (Mono JIT Code) UnityEngine.UIElements.IMGUIContainer:DoOnGUI (UnityEngine.Event,UnityEngine.Matrix4x4,UnityEngine.Rect,bool,UnityEngine.Rect,System.Action,bool)
0x000001e4e323ae3b (Mono JIT Code) UnityEngine.UIElements.IMGUIContainer:HandleIMGUIEvent (UnityEngine.Event,UnityEngine.Matrix4x4,UnityEngine.Rect,System.Action,bool)
0x000001e5c9bd51eb (Mono JIT Code) UnityEngine.UIElements.IMGUIContainer:HandleIMGUIEvent (UnityEngine.Event,System.Action,bool)
0x000001e5c9bd5053 (Mono JIT Code) UnityEngine.UIElements.IMGUIContainer:HandleIMGUIEvent (UnityEngine.Event,bool)
0x000001e5c9bd4b9b (Mono JIT Code) UnityEngine.UIElements.IMGUIContainer:SendEventToIMGUIRaw (UnityEngine.UIElements.EventBase,bool,bool)
0x000001e5c9bd4a83 (Mono JIT Code) UnityEngine.UIElements.IMGUIContainer:SendEventToIMGUI (UnityEngine.UIElements.EventBase,bool,bool)
0x000001e4e2f1022b (Mono JIT Code) UnityEngine.UIElements.IMGUIContainer:HandleEvent (UnityEngine.UIElements.EventBase)
0x000001e4e2f00ed8 (Mono JIT Code) UnityEngine.UIElements.CallbackEventHandler:HandleEventAtTargetPhase (UnityEngine.UIElements.EventBase)
0x000001e5c9bd1f63 (Mono JIT Code) UnityEngine.UIElements.MouseCaptureDispatchingStrategy:DispatchEvent (UnityEngine.UIElements.EventBase,UnityEngine.UIElements.IPanel)
0x000001e4e2edf92e (Mono JIT Code) UnityEngine.UIElements.EventDispatcher:ApplyDispatchingStrategies (UnityEngine.UIElements.EventBase,UnityEngine.UIElements.IPanel,bool)
0x000001e4e2edf4cb (Mono JIT Code) UnityEngine.UIElements.EventDispatcher:ProcessEvent (UnityEngine.UIElements.EventBase,UnityEngine.UIElements.IPanel)
0x000001e4e2f04e6b (Mono JIT Code) UnityEngine.UIElements.EventDispatcher:ProcessEventQueue ()
0x000001e4e2f03643 (Mono JIT Code) UnityEngine.UIElements.EventDispatcher:OpenGate ()
0x000001e4e2f0359b (Mono JIT Code) UnityEngine.UIElements.EventDispatcherGate:Dispose ()
0x000001e4e2edf753 (Mono JIT Code) UnityEngine.UIElements.EventDispatcher:ProcessEvent (UnityEngine.UIElements.EventBase,UnityEngine.UIElements.IPanel)
0x000001e4e2edef93 (Mono JIT Code) UnityEngine.UIElements.EventDispatcher:Dispatch (UnityEngine.UIElements.EventBase,UnityEngine.UIElements.IPanel,UnityEngine.UIElements.DispatchMode)
0x000001e4e2edee2b (Mono JIT Code) UnityEngine.UIElements.BaseVisualElementPanel:SendEvent (UnityEngine.UIElements.EventBase,UnityEngine.UIElements.DispatchMode)
0x000001e4e2ffd17b (Mono JIT Code) UnityEngine.UIElements.UIElementsUtility:DoDispatch (UnityEngine.UIElements.BaseVisualElementPanel)
0x000001e4e2ffcc63 (Mono JIT Code) UnityEngine.UIElements.UIElementsUtility:UnityEngine.UIElements.IUIElementsUtility.ProcessEvent (int,intptr,bool&)
0x000001e4e2ffca5f (Mono JIT Code) UnityEngine.UIElements.UIEventRegistration:ProcessEvent (int,intptr)
0x000001e4e2ffc99b (Mono JIT Code) UnityEngine.UIElements.UIEventRegistration/<>c:<.cctor>b__1_2 (int,intptr)
0x000001e4e2ffc7ed (Mono JIT Code) UnityEngine.GUIUtility:ProcessEvent (int,intptr,bool&)
0x000001e4e2ffc8a6 (Mono JIT Code) (wrapper runtime-invoke) <Module>:runtime_invoke_void_int_intptr_intptr& (object,intptr,intptr,intptr)
0x00007fff76dbe034 (mono-2.0-bdwgc) [mini-runtime.c:3445] mono_jit_runtime_invoke 
0x00007fff76cfe724 (mono-2.0-bdwgc) [object.c:3064] do_runtime_invoke 
0x00007fff76cfe8bc (mono-2.0-bdwgc) [object.c:3111] mono_runtime_invoke 
0x00007ff7b51dabf4 (Unity) scripting_method_invoke
0x00007ff7b51d5814 (Unity) ScriptingInvocation::Invoke
0x00007ff7b51cf135 (Unity) ScriptingInvocation::Invoke<void>
0x00007ff7b52d7d0a (Unity) Scripting::UnityEngine::GUIUtilityProxy::ProcessEvent
0x00007ff7b5d0d9e6 (Unity) GUIView::ProcessRetainedMode
0x00007ff7b62530eb (Unity) GUIView::OnInputEvent
0x00007ff7b5d0d7fa (Unity) GUIView::ProcessInputEvent
0x00007ff7b6254007 (Unity) GUIView::ProcessEventMessages
0x00007ff7b624dc0c (Unity) GUIView::GUIViewWndProc
0x00007ffff32be858 (USER32) CallWindowProcW
0x00007ffff32be299 (USER32) DispatchMessageW
0x00007ff7b6228c43 (Unity) MainMessageLoop
0x00007ff7b622d68b (Unity) WinMain
0x00007ff7b754c1fe (Unity) __scrt_common_main_seh
0x00007ffff4527034 (KERNEL32) BaseThreadInitThunk
0x00007ffff4922651 (ntdll) RtlUserThreadStart

**Comment by aflatau** [comment]
Getting same issues as above - Happens on builds, and its not constant. Sometimes it happens once, then sometimes it happens two or three times in a row. Then it works.. 

Unity - 2021.3.0.f1 

Raw crash log - https://pastebin.com/raw/QN4B6AFD

**Comment by DenisSulyagin** [comment]
Do you use "Steam return" on the mixer? I noticed that it was hanging with a subsequent crash

**Comment by aflatau** [comment]
> Do you use "Steam return" on the mixer? I noticed that it was hanging with a subsequent crash

Yes - I use it for the Reverb

**Comment by DenisSulyagin** [comment]
> > Do you use "Steam return" on the mixer? I noticed that it was hanging with a subsequent crash
> 
> Yes - I use it for the Reverb

I am currently looking for a problem myself. Try not to use anything from Steam Audio on the mixer, and report the results here. Reflections will work. I also noticed that baked reverb doesn't work for me.

**Comment by CodeHatchling** [comment]
I've also noticed crashes after this fix was released. They were a lot more infrequent, but they're still there - in both builds and the editor.
A sore spot seems to be switching levels. There's a much higher chance of a crash occurring there. Perhaps SteamAudio classes are initializing themselves in an unexpected order.

As we're trying to submit a build for Steam approval we'll probably need to disable this component until it is fixed properly. It is kind of ironic that Steam could possibly diapprove our build because of a component they released under their name. :P

In any case, here are crash logs I recovered for the editor in which Steam Audio was the cause:
[crash.zip](https://github.com/ValveSoftware/steam-audio/files/8723051/crash.zip)
[crash2.zip](https://github.com/ValveSoftware/steam-audio/files/8723050/crash2.zip)
